### PR TITLE
Tweak development tips for Ansible operator

### PR DESCRIPTION
**Pull request by tengqm**
**Description of the change:**

- Replace 'k8s Ansible module' with 'Kubernetes Collection for Ansible' where appropriate
- Set code snippet to 'bash' or 'none' properly to avoid confusing highlights
- Update the output from `operator sdk new` to what we actually see today
- Revise the tutorial to create ConfigMap rather than Namespace which doesn't work out of the box for on cluster testing
- Fix the role name spelling, i.e. replace 'Foo' by 'foo' where needed
- Misc text tweaking for readability, e.g. 'inside of an Operator' -> 'inside an Operator'
- Rename "Ansible Operator annotations" to "Annotations for Curstom Resource" to avoid confusion
- Fixed the "Ansible Operator Conditions" section to reflect current status.

**Motivation for the change:**

Fix outdated content and (hopefully) improve the documentation for Ansible Operator developers.

### CODE REVIEW

**Comment by tengqm** [comment]
preview: https://deploy-preview-3236--operator-sdk.netlify.app/docs/ansible/development-tips/

**Comment by tengqm** [comment]
@fabianvf @camilamacedo86 @estroz PTAL? Thanks.

File: website/content/en/docs/ansible/development-tips.md [hunk 1]
```diff
@@ -116,26 +124,25 @@ PLAY [localhost] ***************************************************************
...
-default       	           Active    28d
-kube-public   		   Active    28d
-kube-system   	           Active    28d
-example-memcached          Active    3s
+Check that the ConfigMap was created:
+
+```none
+$ kubectl get configmaps
+NAME                    STATUS    AGE
+example-config          Active    3s
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:140]: all examples are based in the memcached as the sample; https://github.com/operator-framework/operator-sdk-samples/tree/master/ansible/memcached-operator

Can we please use it?

File: website/content/en/docs/ansible/development-tips.md [hunk 2]
```diff
@@ -6,107 +6,115 @@ weight: 12
...
-The easiest way to get started is to install the modules on your local machine
-and test them using a playbook.
+The easiest way to get started is to install the collection on your local
+machine and test them using a playbook.
 
-### Installing the k8s Ansible modules
+### Installing the Kubernetes Collection for Ansible
+
+To install the Kubernetes Collection, one must first install Ansible 2.9+.
+For example, on Fedora/Centos:
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:27]: I understand that it is about https://docs.ansible.com/ansible/latest/modules/k8s_module.html. Am I right? Could we also add the links and clarify that the following steps is specifically about the collection k8s?

File: website/content/en/docs/ansible/development-tips.md [hunk 3]
```diff
@@ -6,107 +6,115 @@ weight: 12
...
-### Testing the k8s Ansible modules locally
+### Testing the Kubernetes Collection locally
 
 Sometimes it is beneficial for a developer to run the Ansible code from their
 local machine as opposed to running/rebuilding the operator each time. To do
 this, initialize a new project:
-```bash
+
+```none
 $ operator-sdk new --type ansible --kind Foo --api-version foo.example.com/v1alpha1 foo-operator
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:66]: Could we replace here for the Memcached one?

File: website/content/en/docs/ansible/development-tips.md (hunk 1 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:140]: Yes, that is an option. However, my feeling is that for this doc we don't want to force the users to clone a different repo (`git clone https://github.com/operator-framework/operator-sdk-samples.git`) to understand the tips.

The previous doc was using namespace creation as a simple example, which is good. I'm changing it to configmap because namespace is a cluster scoped resource which will fail the in-cluster testing, although everything works fine when running `operator-sdk run local`.

File: website/content/en/docs/ansible/development-tips.md (hunk 2 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:27]: No. We have switched over to the [Kubernetes Collection for Ansible](https://galaxy.ansible.com/community/kubernetes) which provides other modules in addition to `k8s_module`. This new link is added in the revised doc.

File: website/content/en/docs/ansible/development-tips.md (hunk 3 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:66]: See above comment.
If you feel strongly about switching to memcached example, I can revise as well.

File: website/content/en/docs/ansible/development-tips.md (hunk 1 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:140]: My suggestion was NOT to use the sample. i do not think that we should link it.
just use the Memcached as an example. See the quick start, it will give the steps to create the Memcached sample. IMO we should stick with it instead of creating here `example-config `

File: website/content/en/docs/ansible/development-tips.md (hunk 2 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:27]: Can we please add the info and the link for it?

File: website/content/en/docs/ansible/development-tips.md (hunk 3 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:66]: See the quick start: https://github.com/operator-framework/operator-sdk/blob/master/website/content/en/docs/ansible/quickstart.md

My suggestion here is: 

```suggestion
$ operator-sdk new memcached-operator --api-version=cache.example.com/v1alpha1 --kind=Memcached --type=ansible
```

Which means explain how the collections work based on in the example that has been provided in the docs already.

File: website/content/en/docs/ansible/development-tips.md (hunk 1 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:140]: I will mark as solved because we have a comment on top about that already.

File: website/content/en/docs/ansible/development-tips.md (hunk 3 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:66]: I think that would be nice we keep using the same example of the quick start. however, the Memcached is not used now and it should not be a block to move forward with your improvements.

File: website/content/en/docs/ansible/development-tips.md [hunk 4]
```diff
@@ -6,107 +6,115 @@ weight: 12
...
 $ cd foo-operator
 $ ansible-galaxy collection install -r requirements.yml
 ```
-Modify `roles/Foo/tasks/main.yml` with desired Ansible logic. For this example
-we will create and delete a namespace with the switch of a variable:
+
+Modify `roles/foo/tasks/main.yml` with desired Ansible logic. For this example
+we will create and delete a ConfigMap based on the value of a variable named
+`state`:
+
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:82]: Why you change the example to use a ConfigMap instead of the namespace?

File: website/content/en/docs/ansible/development-tips.md [hunk 5]
```diff
@@ -432,55 +461,54 @@ Declaring collections in the role meta allows you to invoke the
...
+There are mainly two condition types:
 
-* Successful - if the run has finished and there were no errors, the Ansible
-  Operator will be marked as Successful. It will then wait for the next
-  reconciliation action, either the reconcile period, dependent watches triggers
-  or the resource is updated.
+* `Running` - the Operator is currently running Ansible for reconciliation or
+  if a run has finished with no errors and it is waiting for the next
+  reconciliation action. The next action can be triggered by the reconcile
+  period, dependent watches or an update to the Custome Resource.
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:471]: The : 
```
or
  if a run has finished with no errors and it is waiting for the next
  reconciliation action. The next action can be triggered by the reconcile
  period, dependent watches or an update to the Custome Resource.
```
is not valid for `Running`. 

When the reconciliation finishes it should assume the `Successful` status.

Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:471]: See the code source. The changes made here regards the status are not accurate. 

https://github.com/operator-framework/operator-sdk/blob/ad1d4d1953abf22c336f67198fc23a489c0f4fdb/pkg/ansible/controller/status/utils.go#L22-L30

File: website/content/en/docs/ansible/development-tips.md [hunk 6]
```diff
@@ -6,107 +6,115 @@ weight: 12
...
 
-To install the k8s Ansible modules, one must first install Ansible 2.9+. On
-Fedora/Centos:
 ```bash
 $ sudo dnf install ansible
 ```
 
 In addition to Ansible, a user must install the [OpenShift Restclient
-Python][openshift_restclient_python] package. This can be installed from pip:
+Python](https://github.com/openshift/openshift-restclient-python) package:
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:34]: Please, revert it. The links should be added in the bottom.

File: website/content/en/docs/ansible/development-tips.md [hunk 7]
```diff
@@ -492,17 +520,17 @@ The structure passed to Ansible as extra vars is:
...
+Resource. You can access these metadata  via dot notation in Ansible:
+
 ```yaml
 ---
 - debug:
     msg: "name: {{ meta.name }}, {{ meta.namespace }}"
 ```
 
-[k8s_ansible_module]:https://docs.ansible.com/ansible/2.6/modules/k8s_module.html
-[openshift_restclient_python]:https://github.com/openshift/openshift-restclient-python
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:505]: Revert the link. (line 505)

File: website/content/en/docs/ansible/development-tips.md [hunk 8]
```diff
@@ -492,17 +520,17 @@ The structure passed to Ansible as extra vars is:
    },
 }
 ```
-`message` and `newParameter` are set in the top level as extra variables, and
-`meta` provides the relevant metadata for the Custom Resource as defined in the
-operator. The `meta` fields can be accesses via dot notation in Ansible as so:
+
+The `message` and `newParameter` at the top level are set as extra variables
+themselves. The `meta` key contains some metadata relevant to the Custom
+Resource. You can access these metadata  via dot notation in Ansible:
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:526]:
```suggestion
Resource. The `meta` fields can be accesses via dot notation in Ansible as so:
```

File: website/content/en/docs/ansible/development-tips.md [hunk 9]
```diff
@@ -432,55 +461,54 @@ Declaring collections in the role meta allows you to invoke the
...
-passes along additional variables under the `meta` field for the name of the CR
-and the namespace of the CR.
+## Extra Variables Sent to Ansible
+
+The Operator treats the content you provide in the `spec` section of the CR
+manifest as key-value pairs. It sends them as extra vars  to Ansible. This is
+equivalent to [passing extra vars to
+`ansible-playbook`](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#passing-variables-on-the-command-line).
+The operator also passes some additional key-value pairs using the `meta` key,
+e.g., the name and the namespace of the CR.
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:491]: Can you please revert the changes in the explanation and just add the link reference? 
IMO : The previous explanation is more accurate. 
PS.: The links should be added on the bottom. See the other ones.

File: website/content/en/docs/ansible/development-tips.md [hunk 10]
```diff
@@ -379,29 +412,25 @@ status:
...
-
-This allows the developer to update the `status` from within Ansible
-with any key/value pair as desired. By default, Ansible Operator will
-always include the generic Ansible run output as shown above. If you
-would prefer your application *not* update the status with Ansible
-output and would prefer to track the status manually from your
-application, then simply update the watches file with `manageStatus`:
+An Ansible Operator also allows you to supply custom status values with the
+`k8s_status` Ansible module, which is included in
+[operator_sdk.util](https://galaxy.ansible.com/operator_sdk/util) collection.
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:417]: add the link in the bottom.

File: website/content/en/docs/ansible/development-tips.md [hunk 11]
```diff
@@ -379,29 +412,25 @@ status:
...
+You can update the `status` from within Ansible with any key/value pairs as
+desired. If you want to track the CR status manually from your application,
+you can update the `watches.yaml` file with `manageStatus`, as shown below:
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:420]: Could you revert to the previous description which is more accurate?

File: website/content/en/docs/ansible/development-tips.md (hunk 5 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:471]: Well, I don't agree.
The only place the SuccessfulReason is used (referenced) is here - https://github.com/operator-framework/operator-sdk/blob/master/pkg/ansible/controller/reconcile.go#L386-L392. In that context, the condition *type* is still "Running". 

In fact, when we are talking about condition type, rather than the reason why the operator is in a specific condition type, there are only two choices, not three (https://github.com/operator-framework/operator-sdk/blob/master/pkg/ansible/controller/status/types.go#L88-L96)

```go
// ConditionType - type of condition
type ConditionType string

const (
	// RunningConditionType - condition type of running.
	RunningConditionType ConditionType = "Running"
	// FailureConditionType - condition type of failure.
	FailureConditionType ConditionType = "Failure"
)
```

File: website/content/en/docs/ansible/development-tips.md (hunk 6 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:34]: Okay, I was thinking that maybe the target is only referenced once, we can just put it in place.
For people who work on this doc, scrolling to the bottom to find out the target is a bit inconvenient, unless the target is mentioned several times in the doc.

But anyways, I'll respect the team's convention on this.

File: website/content/en/docs/ansible/development-tips.md (hunk 8 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:526]: I came from a kubernetes docs community where there is a guide line [use active voice](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice).
I'm trying to share some common practices from tech writers, though I am not strongly opinionated on this.

File: website/content/en/docs/ansible/development-tips.md (hunk 9 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:491]: Moved the link to the bottom of the page.
Would love to understand how we can improve this paragraph for accuracy.

File: website/content/en/docs/ansible/development-tips.md (hunk 10 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:417]: Done.

File: website/content/en/docs/ansible/development-tips.md (hunk 11 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:420]: Actually I didn't add or remove any text. A summary of the changes made to the paragraph:

- Changed "Ansible Operator" to "An Ansible Operator" to make it generic and grammatically correct.
- Simplified "you as the developer" to "you" because we are in a doc called development tips after all.
- Dropped "By default, Ansible Operator will always include the generic Ansible run output as shown above." because we already said that in previous paragraph.

This only missing piece (after rereading this several times, :)) is `If you would prefer your application *not* update the status with Ansible output`, which I'll add back in next commit.

File: website/content/en/docs/ansible/development-tips.md (hunk 8 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:526]: Very interesting. I did not know that it has a definition. I agree we should follow up that. Make senses for me. 
Tks for share I will flag it internally as well.

File: website/content/en/docs/ansible/development-tips.md (hunk 5 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:472]: Hi @tenqm,  

If you check the [Memcached Sample](https://github.com/operator-framework/operator-sdk-samples/tree/master/ansible/memcached-operator) when the task finishes the Status will be : 

```
Status:
  Conditions:
    Ansible Result:
      Changed:             0
      Completion:          2020-07-04T09:41:01.05909
      Failures:            0
      Ok:                  2
      Skipped:             0
    Last Transition Time:  2020-07-04T09:40:42Z
    Message:               Awaiting next reconciliation
    Reason:                Successful
    Status:                True
    Type:                  Running
Events:                    <none>
```

See that the type of the conditional is `Running` but the Reason is `Successful` and the doc here is describing the status conditionals reason `Successful` and `Failed`. Because of this, could you please revert your changes and keep it as it was before?

File: website/content/en/docs/ansible/development-tips.md (hunk 9 above)
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md]:
```suggestion
The Operator treats the content you provide in the `spec` section of the CR
manifest as key-value pairs. It sends them as extra vars to Ansible and are managed by the operator. This is
equivalent to how above extra vars are passed into [passing extra vars to
`ansible-playbook`](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#passing-variables-on-the-command-line).
The operator also passes along additional variables under the `meta` field. These variable are key-value pairs, such as for example; the CR name and namespace. 
```

Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:491]: See:

File: website/content/en/docs/ansible/development-tips.md (hunk 5 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:472]: Why revert this? The original text lists three values: `Running`, `Successful` and `Failed`.
The set of condition types:

- `Running`
- `Failure`

The set of condition reasons:

- `Running`: valid when type is `Running`; 
- `Successful`: valid when type is `Running`;
- `Failed`: valid only when type is `Failure`;
- `Unknown`: defined, but not used anywhere, designed for `Failure` condition type.

File: website/content/en/docs/ansible/development-tips.md (hunk 9 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md]: > "and are managed by the operator."

What are managed by the operator? 

> The operator also passes along additional variables under the `meta` field. These variable are key-value pairs, such as for example; the CR name and namespace. 

I don't see any where key-value pairs are called "variables". It is confusing. "key-value pairs" contain both the variables and their values, aren't they?  "under the `meta` field" is also a weird expression ...

File: website/content/en/docs/ansible/development-tips.md (hunk 5 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:472]: > Why revert this? The original text lists three values: Running, Successful and Failed.
Because the previous text was describing the `condition reasons` and you are changing the text. 

I agree that it has a hall for improvements. However, WDYT about we revert it and the working on this area in another PR just to improve the explanation over the status conditions and its reasons to make easier the review and etc. Ideally, each pr should have a close/small scope instead with fewer changes in order to ensure the quality and make the process flow smoothly.

File: website/content/en/docs/ansible/development-tips.md (hunk 9 above)
Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md]: Hi @tengqm, 

I am not native in English, however, shows that your modifications cange a little the understatement and has grammar issues.  I'd suggest revert and then do a PR with a close scope to address your motivations here. Ideally, we should not do a PR with so many different motivations.

Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md]: Em... alright. That makes sense.

File: website/content/en/docs/ansible/development-tips.md (hunk 5 above)
Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:472]: When you examine the status of a resource in Kubernetes, you will first look at the condition.type and then the condition.status. The combination of "type" and "status" gives you an idea whether the resource is operational. If you got an negative combination of the "type" and "status", you will check the "message" and "reason" field. In other words, we are not checking the reasons directly.

If we check the condition type definition again, I'm even thinking whether we should remove the `Failure` condition type because it is redundant. A `Failure` type of condition is equivalent to `Running=False`.

| type | status | reason | note |
|------|-------|--------|------|
| Running | True | Running | all good |
| Running | True | Successful | last reconciliation was successful |
| Running | False | Failed | a known error is caught |
| Running | False | Unknown | something bad happened with unknown reason |
| Running | Unknown | ... | ... |
| Failure | True | ... | ... |
| Failure | False | ... | ... |
| Failure | Unknown | ... | ... |

Anyway, I agree we can leave this untouched at the moment.

File: website/content/en/docs/ansible/development-tips.md [hunk 12]
```diff
@@ -323,44 +331,69 @@ $ kubectl create -f deploy/operator.yaml
...
-...
-- name: ANSIBLE_DEBUG_LOGS
-  value: "True"
-...
+spec:
+  template:
+    spec:
+      containers:
+        - name: foo-operator
+          env:
```
Comment (camilamacedo86) [inline website/content/en/docs/ansible/development-tips.md:382]:
```suggestion
          ...
          env:
```

Just to make clear that has more code between the name and the env.

Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:382]: That is doable, but I have already marked this code snippet as YAML. Adding "..." may break the syntax, I'm afraid.
This snippet is obviously not a complete one...

Anyway, if you insist, I can send another commit.

Comment (camilamacedo86) [reply website/content/en/docs/ansible/development-tips.md:383]: It is the default standard that we work with to illustrate has more code and do not lead misunderstandings. you can check the other docs. Could you please just commit this suggestion? Then, for me it would be ok but before merge requires more one approval which means a second reviewer too.

Comment (tengqm) [reply website/content/en/docs/ansible/development-tips.md:383]: see 09c91b7

**Comment by camilamacedo86** [comment]
Hi @tengqm, 

Could you just rebase with the master (a small change that requires that)

**Comment by tengqm** [comment]
@camilamacedo86 Done.

**Comment by camilamacedo86** [comment]
Hi @tengqm, 

we need to have 2 lgtms to go forward. And now it needs to rebase .. Could you please rebase it with the master and then ping the pr in the slack channel (https://kubernetes.slack.com/archives/C017UU45SHL) asking for reviews as well?
